pkg/packagekit/kpatch.jsx

```diff
@@ -0,0 +1,376 @@
+
+        cockpit.spawn(["uname", "-r"])
+                .then(data => {
+                    const fields = data.split(".");
+                    const sanitized_kernel_ver = fields.slice(0, fields.length - 2)
+                            .join("_");
+                    const patch_name = "kpatch-patch-" + sanitized_kernel_ver;
+                    check_missing_packages([patch_name])
+                            .then(d => {
+                                this.setState((state, _) => {
+                                    const installed = (d.unavailable_names || []).length === 0 && (d.missing_names || []).length === 0;
+                                    const unavailable = (d.unavailable_names || []).length > 0;
+                                    const current = state.enabled && (installed || unavailable);
+                                    return ({
+                                        patchName: patch_name,
+                                        patchInstalled: installed,
+                                        patchUnavailable: unavailable,
+                                        justCurrent: current && !state.auto,
+                                        applyCheckbox: current,
+                                        kernelName: data,
+                                    });
+                                });
+                            })
+                            .catch(() => true);
+                })
+                .catch(() => true);
+    }
+
+    handleInstall() {
+        this.setState({ updating: true });
+        install_dialog(this.state.missing)
+                .then(() => {
+                    this.setState({ missing: [], updating: false });
+                })
+                .catch(() => this.setState({ updating: false }));
+    }
+
+    onClose() {
+        this.setState((state, _) => {
+            const current = state.enabled && (state.patchInstalled || state.patchUnavailable);
+            return ({
+                justCurrent: current && !state.auto,
+                applyCheckbox: current,
+                showModal: false,
+                error: "",
+            });
+        });
+    }
+
+    handleChange() {
+        this.setState({ updating: true });
+
+        if (this.state.applyCheckbox) {
+            let install = Promise.resolve(true);
+            if (this.state.justCurrent) {
+                install = new Promise((resolve, reject) => {
+                    cockpit.spawn(["dnf", "-y", "kpatch", "manual"], { superuser: "require", err: "message" })
+                            .then(() => {
+                                if (!this.state.patchUnavailable && !this.state.patchInstalled)
+                                    cockpit.spawn(["dnf", "-y", "install", this.state.patchName], { superuser: "require", err: "message" }).then(resolve)
+                                            .catch(reject);
```

Comment:
Do we really need this `new Promise()` wrapper? simply chaining the dnf kpatch and dnf install promises should DTRT?

Reply:
in `then()` there might or might not be a next promise - see the if/else. Honestly not sure how it would behave in the else-case, but this approach seems more clear to me.